src/udev.c

```diff
@@ -224,6 +235,119 @@ udev_find_more_about_optical(struct udev_device *udev_device,  device_t *device,
   udev_enumerate_unref(enumerate);
 }
 
+static int
+is_ethernet_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == COMMUNICATIONS_CLASS &&
+    interface->bInterfaceSubClass == ETHERNET_NETWORKING_SUBCLASS);
+}
+
+static int
+is_bluetooth_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == WIRELESS_CLASS &&
+    interface->bInterfaceSubClass == RADIO_FREQUENCY_SUBCLASS &&
+    interface->bInterfaceProtocol == BLUETOOTH_PROTOCOL);
+}
+
+/**
+ * Identify vendor-specific interfaces with endpoints that support
+ * bulk data transfers both in and out with a packet size of at least
+ * TYPICAL_NIC_PACKET_SIZE (512). Though all the USB NICs that were
+ * tested had such an endpoint and none of the other USB devices did,
+ * It is possible that checking for NICs in this manner will result
+ * in false positives or negatives.
+ */
+static int
+is_unknown_bulk_transfer_interface(struct usb_interface_descriptor *interface)
+{
+  int i;
+  int has_bulk_in = 0;
+  int has_bulk_out = 0;
+
+  if (interface->bInterfaceClass == VENDOR_SPECIFIC_CLASS)
+    for (i = 0; i < interface->bNumEndpoints; i++)
+    {
+      struct usb_endpoint_descriptor *endpoint = &interface->endpoint[i];
+
+      if (endpoint->bmAttributes == BMATTRIBUTES_BULK && endpoint->wMaxPacketSize >= TYPICAL_NIC_PACKET_SIZE)
+      {
+        if (endpoint->bEndpointAddress & BENDPOINTADDRESS_IN)
+          has_bulk_in = 1;
+        else
+          has_bulk_out = 1;
+      }
+    }
+  return (has_bulk_in && has_bulk_out);
```

Comment:
Minor: this could be optimized by returning/breaking as soon as both int (why not bool btw?) are true.